src/services/refactors/convertStringOrTemplateLiteral.ts

```diff
@@ -101,32 +101,31 @@ namespace ts.refactor.convertStringOrTemplateLiteral {
             }
         });
 
-        return container || expr;
+        return (container || expr) as Expression;
     }
 
-    function isStringConcatenationValid(node: Node): boolean {
-        const { containsString, areOperatorsValid } = treeToArray(node);
-        return containsString && areOperatorsValid;
-    }
-
-    function treeToArray(current: Node): { nodes: Expression[], operators: Token<BinaryOperator>[], containsString: boolean, areOperatorsValid: boolean} {
-        if (isBinaryExpression(current)) {
-            const { nodes, operators, containsString: leftHasString, areOperatorsValid: leftOperatorValid } = treeToArray(current.left);
+    function treeToArray(current: Expression) {
+        const loop = (current: Node): { nodes: Expression[], operators: Token<BinaryOperator>[], hasString: boolean, validOps: boolean} => {
+            if (!isBinaryExpression(current)) {
+                return { nodes: [current as Expression], operators: [], validOps: true,
+                         hasString: isStringLiteral(current) || isNoSubstitutionTemplateLiteral(current) };
+            }
+            const { nodes, operators, hasString: leftHasString, validOps: leftOperatorValid } = loop(current.left);
 
-            if (!leftHasString && !isStringLiteral(current.right) && !isTemplateExpression(current.right)) {
-                return { nodes: [current], operators: [], containsString: false, areOperatorsValid: true };
+            if (!(leftHasString || isStringLiteral(current.right) || isTemplateExpression(current.right))) {
+                return { nodes: [current], operators: [], hasString: false, validOps: true };

```

Comment:
If you really don't like the old name I might suggest `validOps` -> `validOperators` instead so it's more apparent what this is referring to.

Reply:
I actually prefer the shorter names, and did that because it is used only in the local function, so I'd prefer changing `hasString` to `hasStr` to make it more uniform; but short variable names are always a subjective issue, so I just applied your suggestion instead...